### PR TITLE
Adding ArcAssay/ArcStudy API

### PR DESCRIPTION
**Pull request by Freymaurer**
- Hit >300 unit tests 🎉 ✅ 
- JS tests not yet run ❌

### CODE REVIEW

**Comment by Freymaurer** [comment]
@HLWeil please have a look!

File: src/ISA/ISA/ArcTypes/ArcAssay.fs [hunk 1]
```diff
@@ -2,6 +2,64 @@
 
 open Fable.Core
 
+module ArcAssayAux =
+
+    let getNextAutoGeneratedTableName (existingNames: seq<string>) =
```
Comment (HLWeil) [inline src/ISA/ISA/ArcTypes/ArcAssay.fs:7]: I would enforce users to set table names themselves when using this API.

File: src/ISA/ISA/ArcTypes/ArcTable.fs [hunk 2]
```diff
@@ -400,4 +422,17 @@ type ArcTable =
...
+        match other with
+        | :? ArcTable as table -> 
+            let sameName = this.Name = table.Name
+            let sameHeaders = Seq.forall2 (fun x y -> x = y) this.Headers table.Headers
+            let sameBodyCells = Seq.forall2 (fun x y -> x = y) this.Headers table.Headers
+            sameName && sameHeaders && sameBodyCells
+        | _ -> false
+
+    // it's good practice to ensure that this behaves using the same fields as Equals does:
+    override this.GetHashCode () = this.GetHashCode()
```
Comment (HLWeil) [inline src/ISA/ISA/ArcTypes/ArcTable.fs:438]: rip

File: src/ISA/ISA/ArcTypes/ArcAssay.fs [hunk 3]
```diff
@@ -24,26 +82,349 @@ type ArcAssay =
...
     [<NamedParams>]
-    static member create (?ID : URI, ?FileName : string, ?MeasurementType : OntologyAnnotation, ?TechnologyType : OntologyAnnotation, ?TechnologyPlatform : string, ?Sheets : ArcTable list, ?Performers : Person list, ?Comments : Comment list) = 
+    static member create (?ID : URI, ?FileName : string, ?MeasurementType : OntologyAnnotation, ?TechnologyType : OntologyAnnotation, ?TechnologyPlatform : string, ?Sheets : ResizeArray<ArcTable>, ?Performers : Person list, ?Comments : Comment list) = 
+        let Sheets = ResizeArray()
         ArcAssay.make ID FileName MeasurementType TechnologyType TechnologyPlatform Sheets Performers Comments
 
+
+    // - Table API - //
+    // remark should this return ArcTable?
+    member this.AddTable(?table:ArcTable, ?index: int) =
```
Comment (HLWeil) [inline src/ISA/ISA/ArcTypes/ArcAssay.fs:114]:
```fsharp
member this.AddTable(table:ArcTable, ?index: int) : unit

member this.InitTable(tableName : string, ?index : int) : ArcTable
```

Comment (HLWeil) [reply src/ISA/ISA/ArcTypes/ArcAssay.fs:114]: Maybe something like this?

File: src/ISA/ISA/ArcTypes/ArcAssay.fs [hunk 4]
```diff
@@ -24,26 +82,349 @@ type ArcAssay =
...
+        this.Tables.[index]
+
+    static member getTableAt(index:int) : ArcAssay -> ArcTable =
+        fun (assay:ArcAssay) ->
+            let newAssay = assay.Copy()
+            newAssay.GetTableAt(index)
+
+    // - Table API - //
+    member this.GetTable(name: string) : ArcTable =
+        tryByTableName name this.Tables
```
Comment (HLWeil) [inline src/ISA/ISA/ArcTypes/ArcAssay.fs:141]: rename

File: src/ISA/ISA/ArcTypes/ArcAssay.fs [hunk 5]
```diff
@@ -24,26 +82,349 @@ type ArcAssay =
...
+
+    // - Table API - //
+    member this.GetTableAt(index:int) : ArcTable =
+        SanityChecks.validateSheetIndex index false this.Tables
+        this.Tables.[index]
+
+    static member getTableAt(index:int) : ArcAssay -> ArcTable =
+        fun (assay:ArcAssay) ->
+            let newAssay = assay.Copy()
+            newAssay.GetTableAt(index)
```
Comment (HLWeil) [inline src/ISA/ISA/ArcTypes/ArcAssay.fs:137]: - add comment to make clear that the user receives a copied object
- copy table, not assay

**Review by HLWeil — CHANGES_REQUESTED** [review]
<3

**Comment by Freymaurer** [comment]
Renamed "Set" CRUD api functions to "Updated"; `SetColumn` --> `UpdateColumn`
Renamed `UpdateTableAt` functions on ArcAssay and ArcStudy to `MapTableAt`

```fsharp
member this.MapTableAt(index: int, updateFun: ArcTable -> unit) = ArcTables(this.Tables).MapTableAt(index, updateFun)
```

**Comment by HLWeil** [comment]
Great work! 

I will merge this, so I can start using these functions in the FsSpreadsheet IO. If we decide on additional changes feel free to open up a new PR.

**Pull request merged**